pkg/etcd/snapshot_handler.go

```diff
@@ -0,0 +1,205 @@
+		util.SendError(err, rw, req, http.StatusInternalServerError)
+		return
+	}
+	sendSnapshotList(rw, req, sf)
+}
+
+func (e *ETCD) handlePrune(rw http.ResponseWriter, req *http.Request) {
+	deleted, err := e.PruneSnapshots(req.Context())
+	if err != nil {
+		util.SendError(err, rw, req, http.StatusInternalServerError)
+		return
+	}
+	sr := &managed.SnapshotResult{
+		Deleted: deleted,
+	}
+	sendSnapshotResponse(rw, req, sr)
+}
+
+func (e *ETCD) handleDelete(rw http.ResponseWriter, req *http.Request, snapshots []string) {
+	deleted, err := e.DeleteSnapshots(req.Context(), snapshots)
+	if err != nil {
+		util.SendError(err, rw, req, http.StatusInternalServerError)
+		return
+	}
+	sr := &managed.SnapshotResult{
+		Deleted: deleted,
+	}
+	sendSnapshotResponse(rw, req, sr)
+}
+
+// withRequest returns a modified ETCD struct that is overriden
+// with etcd snapshot config from the snapshot request.
+func (e *ETCD) withRequest(sr *SnapshotRequest) *ETCD {
+	re := &ETCD{
+		client: e.client,
+		config: &config.Control{
+			CriticalControlArgs:   e.config.CriticalControlArgs,
+			Runtime:               e.config.Runtime,
+			DataDir:               e.config.DataDir,
+			Datastore:             e.config.Datastore,
+			EtcdSnapshotCompress:  e.config.EtcdSnapshotCompress,
+			EtcdSnapshotName:      e.config.EtcdSnapshotName,
+			EtcdSnapshotRetention: e.config.EtcdSnapshotRetention,
+		},
+		name:        e.name,
+		address:     e.address,
+		cron:        e.cron,
+		cancel:      e.cancel,
+		snapshotSem: e.snapshotSem,
+	}
+	if len(sr.Name) > 0 {
+		re.config.EtcdSnapshotName = sr.Name[0]
+	}
+	if sr.Compress != nil {
+		re.config.EtcdSnapshotCompress = *sr.Compress
+	}
+	if sr.Dir != nil {
+		re.config.EtcdSnapshotDir = *sr.Dir
+	}
+	if sr.Retention != nil {
+		re.config.EtcdSnapshotRetention = *sr.Retention
+	}
+	if sr.S3 != nil {
+		re.config.EtcdS3 = true
+		re.config.EtcdS3BucketName = sr.S3.Bucket
+		re.config.EtcdS3AccessKey = sr.S3.AccessKey
+		re.config.EtcdS3SecretKey = sr.S3.SecretKey
+		re.config.EtcdS3EndpointCA = sr.S3.EndpointCA
+		re.config.EtcdS3SkipSSLVerify = sr.S3.SkipSSLVerify
+		re.config.EtcdS3Insecure = sr.S3.Insecure
+		re.config.EtcdS3Region = sr.S3.Region
+		re.config.EtcdS3Timeout = sr.S3.Timeout.Duration
+	}
+	return re
+}
+
+// getSnapshotRequest unmarshalls the snapshot operation request from a client.
+func getSnapshotRequest(req *http.Request) (*SnapshotRequest, error) {
+	switch req.Method {
+	case http.MethodGet:
+		return &SnapshotRequest{
+			Operation: SnapshotOperationList,
+			ctx:       req.Context(),
+		}, nil
+	case http.MethodPost:
+		sr := &SnapshotRequest{}
+		b, err := io.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(b, &sr); err != nil {
+			return nil, err
+		}
+		sr.ctx = req.Context()
+		return sr, nil
+	default:
+		return nil, http.ErrNotSupported
+	}
+}
+
+func sendSnapshotResponse(rw http.ResponseWriter, req *http.Request, sr *managed.SnapshotResult) {
```

Comment:
I assume we are sending the full error back to the client on failure, we don't try and hide it like secrets-encryption?

Reply:
Many errors are probably just returned as-is at the moment. Let me take a look at whether or not that exposes anything.

Reply:
It should be fine, but if you want to obfuscate you may find https://github.com/k3s-io/k3s/blob/master/pkg/server/secrets-encrypt.go#L471 helpful as it makes grepping logs alot easier for the user.

Reply:
I refactored this a bit. Errors that indicate problems with the input from the user are still passed through with minimal information, for example:

```console
root@k3s-server-1:~# k3s etcd-snapshot save --dir /foo
FATA[0000] see server log for details: Internal error occurred: failed to get etcd-snapshot-dir: stat /foo: no such file or directory

root@k3s-server-1:~# k3s etcd-snapshot save --etcd-s3
FATA[0000] see server log for details: s3 bucket name was not set

root@k3s-server-1:~# k3s etcd-snapshot save --etcd-s3 --etcd-s3-bucket foo
FATA[0001] see server log for details: 401 Unauthorized
```